src/game.cpp

```diff
@@ -578,54 +629,193 @@ void test_game() //!OCLINT tests may be many
     assert(losing_player_size_after < losing_player_size_before);
   }
 
-  //#define FIX_ISSUE_381
+  #define FIX_ISSUE_381
   #ifdef FIX_ISSUE_381
-  ///A player can become invulnerable
+  /// A player can become passive
+  {
+    game g;
+
+    assert(is_active(g.get_player(0)));
+    g.set_player_state_passive(g.get_player(0));
+    assert(is_passive(g.get_player(0)));
+  }
+
+  /// A player can become passive, using player.set_state_passive()
   {
     game g;
 
     assert(is_active(g.get_player(0)));
-    become_invulnerable(g.get_player(0));
-    assert(is_invulnerable(g.get_player(0)));
+    g.get_player(0).set_state_passive();
+    assert(is_passive(g.get_player(0)));
+  }
+
+  /// A player can become non-passive again
+  {
+    game g;
+
+    assert(is_active(g.get_player(0)));
+    g.set_player_state_passive(g.get_player(0));
+    g.remove_player_state_passive(g.get_player(0));
+    assert(!is_passive(g.get_player(0)));
+  }
+
+  /// A player can become non-passive again, using player.remove_state_passive()
+  {
+    game g;
 
+    assert(is_active(g.get_player(0)));
+    g.get_player(0).set_state_passive();
+    g.get_player(0).remove_state_passive();
+    assert(!is_passive(g.get_player(0)));
   }
   #endif
```

Comment:
I don't understand why you changed the tests here. I see that you're being more exhaustive, systematically asserting that all setters exist. I don't see the necessity for this; the short test I had written was enough to reach the goal of the issue ("A player can become passive/invulnerable"). Being able to turn it back is not a necessity to reach this.
I believe (I'm not sure, can't read it from your commit history) what you are doing here is characteristic of unit-testing, that is writing tests, typically after the code, to ensure that the concerned bit of code keeps working as intended. This is ok, but we prefer test-driven development. The [difference ](https://methodpoet.com/unit-testing-vs-tdd/) is that in TDD, we write the test *before* the code and even before thinking of the implementation. The test states our intended goal, and will pass when this goal is reached. Then we seek to implement a solution.

This is good work, your tests are functional and do satisfy the end goal, I just want to raise the point that I believe this is excessive, maybe premature testing.

Reply:
@TheoPannetier Thanks for pointing out this, it was that I feel like the tests I had written beforehand were not enough to detect all the changes in the future that might be done intentionally or unintentionally by me or by others, I am sometimes too obsessive that I force myself to add a lot of probably excessive or premature tests. 😢

Reply:
>  I am sometimes too obsessive

I definitely can't blame you for *caring* about the future of the game! I confirm I was too harsh with my initial comment, and I'd like to apologise for this. 

These tests are indeed a bit premature at this stage, but they don't remove anything or add anything dangerous to the game AFAICS and may pave the way for the next steps with the passive state. Take the above as advice on how to better design tests, but what you're doing is an important step too.

Reply:
> > I am sometimes too obsessive
> 
> I definitely can't blame you for _caring_ about the future of the game! I confirm I was too harsh with my initial comment, and I'd like to apologise for this.
> 
> These tests are indeed a bit premature at this stage, but they don't remove anything or add anything dangerous to the game AFAICS and may pave the way for the next steps with the passive state. Take the above as advice on how to better design tests, but what you're doing is an important step too.

This mechanism really involves a lot of worth-thinking issues, I've been a bit busy with my paper recently, I would find a time to rethink about it, and try to improve this PR